### PR TITLE
Contracts version without clearing

### PR DESCRIPTION
**Pull request by hatskier**

### CODE REVIEW

File: contracts/PriceFeed.sol [hunk 1]
```diff
@@ -32,25 +32,24 @@ contract PriceFeed is IPriceFeed, PriceModel, Ownable {
...
         require(isSigner(signer), "Unauthorized price data signer");
-        console.log("Signer OK");
-        console.log("BT: ", block.timestamp * 1000);
-        console.log("PT: ", priceData.timestamp);        
-        require(block.timestamp * 1000 > priceData.timestamp - 15000, "Price data timestamp cannot be from the future");
-        require(block.timestamp * 1000 < priceData.timestamp || block.timestamp * 1000 - priceData.timestamp < maxPriceDelay * 1000, "Price data timestamp too old");
-        console.log("Timestamp OK");
+        // TODO: check the problem with prices on Kovan
+        require(blockTimestampMillseconds > priceData.timestamp - 15000, "Price data timestamp cannot be from the future");
+        require(blockTimestampMillseconds - priceData.timestamp < maxPriceDelay * 1000, "Price data timestamp too old");
```
Comment (jakub-wojciechowski) [inline contracts/PriceFeed.sol:40]: Lets also keep max price delay in milliseconds

File: contracts/PriceFeed.sol [hunk 2]
```diff
@@ -32,25 +32,24 @@ contract PriceFeed is IPriceFeed, PriceModel, Ownable {
...
+
         require(isSigner(signer), "Unauthorized price data signer");
-        console.log("Signer OK");
-        console.log("BT: ", block.timestamp * 1000);
-        console.log("PT: ", priceData.timestamp);        
-        require(block.timestamp * 1000 > priceData.timestamp - 15000, "Price data timestamp cannot be from the future");
-        require(block.timestamp * 1000 < priceData.timestamp || block.timestamp * 1000 - priceData.timestamp < maxPriceDelay * 1000, "Price data timestamp too old");
-        console.log("Timestamp OK");
+        // TODO: check the problem with prices on Kovan
+        require(blockTimestampMillseconds > priceData.timestamp - 15000, "Price data timestamp cannot be from the future");
```
Comment (jakub-wojciechowski) [inline contracts/PriceFeed.sol:39]: Let's extract 15000 to a const,
if the problem on Kovan an mainnet perrsist we could have it as a config parameter

File: contracts/PriceFeedWithoutClearing.sol [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+pragma solidity ^0.8.2;
+pragma experimental ABIEncoderV2;
+
+import './PriceVerifier.sol';
+import './IPriceFeed.sol';
+import 'hardhat/console.sol';
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+contract PriceFeedWithoutClearing is IPriceFeed, PriceModel, Ownable {
```
Comment (jakub-wojciechowski) [inline contracts/PriceFeedWithoutClearing.sol:11]: How about calling it just PrriceFeed and extending by adding one method (+ one line to set current setter) to have the clearing version.

File: contracts/RedstoneCoreProxy.sol [hunk 4]
```diff
@@ -22,50 +22,44 @@ abstract contract RedstoneCoreProxy is Proxy {
      */
     function _delegate(address implementation) internal override {
 
-        //Check if transaction contains Limestone marker
+        // Check if transaction contains Redstone marker
         bool isTxWithPricing = false;
         if (msg.data.length > 32) {
             isTxWithPricing = msg.data.toBytes32(msg.data.length - 32) == MARKER;
```
Comment (jakub-wojciechowski) [inline contracts/RedstoneCoreProxy.sol:28]: Let's intrroduce 2 constants:
MARKER_SLOT_SIZE = 32
LEN_SLOT_SIZE = 2
To improve readibility

File: contracts/RedstoneProxy.sol [hunk 5]
```diff
@@ -46,4 +49,19 @@ contract RedstoneProxy is RedstoneCoreProxy, ERC1967Upgrade {
...
+        require(_priceFeedAddress != address(0), "Price feed address cannot be empty");
+        
+        assert(_IMPLEMENTATION_SLOT == bytes32(uint256(keccak256("eip1967.proxy.implementation")) - 1));
+        StorageSlot.getAddressSlot(_PRICE_FEED_SLOT).value = _priceFeedAddress;
+    }
+
+    /**
+     * @dev Sets the price feed address (internal function)
+     */
+    function setPriceFeed(address _priceFeedAddress) external onlyOwner {
```
Comment (jakub-wojciechowski) [inline contracts/RedstoneProxy.sol:64]: Shouldn't that be managed by admin account rather by owner (and without introducing Ownable), in a pattern similar to :
https://github.com/redstone-finance/redstone-flash-storage/blob/master/contracts/RedstoneUpgradeableProxy.sol#L97

File: test/contracts/DefiWithRedstonePriceFeedWithoutClearing.test.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+        defi = (await Defi.deploy()) as MockDefi;
+
+        // console.log("Defi address: " + defi.address);
+        const proxy = await Proxy.deploy(defi.address, priceFeed.address, admin.address, []);
+
+        defi = (await Defi.attach(proxy.address)) as MockDefi;
+        await defi.initialize(priceFeed.address);
+
+        //defi = defi.connect(signer);
```
Comment (jakub-wojciechowski) [inline test/contracts/DefiWithRedstonePriceFeedWithoutClearing.test.ts:60]: We can remove this line

File: test/contracts/DefiWithRedstonePriceFeedWithoutClearing.test.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+            defi.deposit(toBytes32("FB"), 1),
+            loadApiPrices(),
+        ]);
+
+    });
+
+
+    it("Should inject correct prices from API single", async function() {
+
+        // const apiPrices = await redstone.getAllPrices({provider:"redstone-stocks"});
```
Comment (jakub-wojciechowski) [inline test/contracts/DefiWithRedstonePriceFeedWithoutClearing.test.ts:101]: Let's remove it

File: utils/contract-wrapper.ts [hunk 8]
```diff
@@ -1,47 +1,43 @@
...
     
-    let setPriceData = setPriceTx.data ? setPriceTx.data.substr(2) : "";
+    const setPriceData = remove0xFromHexString(setPriceTx.data, {
+      allowFalsyValues: true,
+    });
 
-    let clearPriceTx = await priceFeed.populateTransaction.clearPrices(priceData);
-    let clearPricePrefix = clearPriceTx.data ? clearPriceTx.data.substr(2,8) : "";
+    let clearPriceTx = await priceFeed.populateTransaction.clearPrices(priceData); // priceData may have any value, we don't use it
+    let clearPricePrefix = clearPriceTx.data ? clearPriceTx.data.substr(2,8) : ""; // we get 4 bytes from 2nd to 5th (skipping 1st (0x)) - this is the function signature clearPrices
```
Comment (jakub-wojciechowski) [inline utils/contract-wrapper.ts:31]: The comment could be betterr:
We skip two first characters ("0x") and get 8 next (4 bytes of signature encoded as HEX)
2nd and 5th could be misleading as we see (2,8)

File: utils/contract-wrapper.ts [hunk 9]
```diff
@@ -1,47 +1,43 @@
...
 
-    //Add priceDataLen info
-    let priceDataLen = setPriceData.length/2;
-    console.log("Price data len: " + priceDataLen);
+    // Add priceDataLen info
+    const priceDataLen = countBytesInHexString(setPriceData);
 
-    return clearPricePrefix + setPriceData + priceDataLen.toString(16).padStart(4, "0");
+    // Template of data that we add to the end of tx is below
+    // [CLEAR_PRICE_FUNCTION_SIGNATURE (4 bytes)] + [SET_PRICE_FUNCTION_WITH_ARGUMENTS (Variable bytes size, 4 bytes for signature + variable length bytes for arguments. It is callable)] + [SET_PRICE_FUNCTION_WITH_ARGS_BYTE_LENGTH (2 bytes)]
```
Comment (jakub-wojciechowski) [inline utils/contract-wrapper.ts:37]: How about adding this as a comment to show structurre:
[SIG_CLEAR|4][SIG_SET|4][DATA|...][DATA_LEN|2][MARKER|32]
and then explaining each part below

**Review by jakub-wojciechowski — CHANGES_REQUESTED** [review]

File: contracts/PriceFeedWithoutClearing.sol (hunk 3 above)
Comment (hatskier) [reply contracts/PriceFeedWithoutClearing.sol]: Great idea, done

**Pull request merged**